# VARSHA SHAJAN


# E-MAIL SPAM DETECTION

Importing the necessary libraries

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

# Reading the dataset

In [2]:
df=pd.read_csv(r"C:\Users\User\Desktop\spam.csv",encoding=('ISO-8859-1'),low_memory=False)

In [3]:
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

In [4]:
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [5]:
df.tail()

v1                                                 v2 Unnamed: 2  \
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN

# Detailing the dataset

In [6]:
df.describe()

v1                      v2  \
count   5572                    5572   
unique     2                    5169   
top      ham  Sorry, I'll call later   
freq    4825                      30   

                                               Unnamed: 2  \
count                                                  50   
unique                                                 43   
top      bt not his girlfrnd... G o o d n i g h t . . .@"   
freq                                                    3   

                   Unnamed: 3 Unnamed: 4  
count                      12          6  
unique                     10          5  
top      MK17 92H. 450Ppw 16"    GNT:-)"  
freq                        2          2

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


In [8]:
df.shape

(5572, 5)

The datset contain 5572 rows and 5 columns.The function describe gives us the details about the dataset.


# Re-naming the dataset

In [9]:
df.rename(columns={'v1':'ham or spam'},inplace=True)

In [10]:
df.rename(columns={'v2':'mail'},inplace=True)

# Cleaning the dataset

In [11]:
df.isnull().sum()

ham or spam       0
mail              0
Unnamed: 2     5522
Unnamed: 3     5560
Unnamed: 4     5566
dtype: int64

In [12]:
df=df.drop(['Unnamed: 2','Unnamed: 3','Unnamed: 4'],axis=1)

In [13]:
df

ham or spam                                               mail
0            ham  Go until jurong point, crazy.. Available only ...
1            ham                      Ok lar... Joking wif u oni...
2           spam  Free entry in 2 a wkly comp to win FA Cup fina...
3            ham  U dun say so early hor... U c already then say...
4            ham  Nah I don't think he goes to usf, he lives aro...
...          ...                                                ...
5567        spam  This is the 2nd time we have tried 2 contact u...
5568         ham              Will Ì_ b going to esplanade fr home?
5569         ham  Pity, * was in mood for that. So...any other s...
5570         ham  The guy did some bitching but I acted like i'd...
5571         ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [20]:
df.duplicated().sum()

403

In [23]:
df=df.drop_duplicates()

All the null values is been removed from the dataset. It also contain 403 dupliacted values.All the noise is been removed

# Modeling


In [27]:
from sklearn.preprocessing import LabelEncoder

In [28]:
encoder=LabelEncoder()

In [29]:
df['ham or spam']=encoder.fit_transform(df['ham or spam'])
df

C:\Users\User\AppData\Local\Temp\ipykernel_2460\3341669943.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['ham or spam']=encoder.fit_transform(df['ham or spam'])


ham or spam                                               mail
0               0  Go until jurong point, crazy.. Available only ...
1               0                      Ok lar... Joking wif u oni...
2               1  Free entry in 2 a wkly comp to win FA Cup fina...
3               0  U dun say so early hor... U c already then say...
4               0  Nah I don't think he goes to usf, he lives aro...
...           ...                                                ...
5567            1  This is the 2nd time we have tried 2 contact u...
5568            0              Will Ì_ b going to esplanade fr home?
5569            0  Pity, * was in mood for that. So...any other s...
5570            0  The guy did some bitching but I acted like i'd...
5571            0                         Rofl. Its true to its name

[5169 rows x 2 columns]

# Splitting the dataset

In [30]:
x=df['mail']
y=df['ham or spam']

In [33]:
from sklearn.model_selection import train_test_split

In [34]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=1)

In [35]:
x_train

2488    Aiyah e rain like quite big leh. If drizzling ...
1754    How is your schedule next week? I am out of to...
3093    Have you emigrated or something? Ok maybe 5.30...
3984             Whatever, juliana. Do whatever you want.
5371               Lol, oh you got a friend for the dog ?
                              ...                        
2904    Ha. You don‰Û÷t know either. I did a a clever ...
925          K, wait chikku..il send aftr  &lt;#&gt; mins
4249                              How much for an eighth?
239     U 447801259231 have a secret admirer who is lo...
5560                    Anything lor. Juz both of us lor.
Name: mail, Length: 3618, dtype: object

In [36]:
y_train

2488    0
1754    0
3093    0
3984    0
5371    0
       ..
2904    0
925     0
4249    0
239     1
5560    0
Name: ham or spam, Length: 3618, dtype: int32

By splitting the datset 70% is used for training the dataset and remaining 30% is used for testing.

# Naive bayes classifier

# Vectorizing

In [40]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [41]:
tfidf=TfidfVectorizer()

In [42]:
x_train_count = tfidf.fit_transform(x_train.values)
x_train_count.toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

# Fitting the model

In [43]:
from sklearn.naive_bayes import MultinomialNB

In [44]:
model=MultinomialNB()
model.fit(x_train_count,y_train)

MultinomialNB()

# Evaluating the model

In [45]:
x_test_count = tfidf.transform(x_test)
model.score(x_test_count,y_test)

0.9445519019987105

# THANKYOU